In [47]:
## Como en el anterior Sprint, no exploré los datos lo suficiente, lo he querido hacer con este sprint. 

In [48]:
import pandas as pd

df = pd.read_csv('housing data.csv',header=None)
df

0     1      2   3      4      5     6       7   8      9     10  \
0    0.00632  18.0   2.31   0  0.538  6.575  65.2  4.0900   1  296.0  15.3   
1    0.02731   0.0   7.07   0  0.469  6.421  78.9  4.9671   2  242.0  17.8   
2    0.02729   0.0   7.07   0  0.469  7.185  61.1  4.9671   2  242.0  17.8   
3    0.03237   0.0   2.18   0  0.458  6.998  45.8  6.0622   3  222.0  18.7   
4    0.06905   0.0   2.18   0  0.458  7.147  54.2  6.0622   3  222.0  18.7   
..       ...   ...    ...  ..    ...    ...   ...     ...  ..    ...   ...   
501  0.06263   0.0  11.93   0  0.573  6.593  69.1  2.4786   1  273.0  21.0   
502  0.04527   0.0  11.93   0  0.573  6.120  76.7  2.2875   1  273.0  21.0   
503  0.06076   0.0  11.93   0  0.573  6.976  91.0  2.1675   1  273.0  21.0   
504  0.10959   0.0  11.93   0  0.573  6.794  89.3  2.3889   1  273.0  21.0   
505  0.04741   0.0  11.93   0  0.573  6.030  80.8  2.5050   1  273.0  21.0   

         11    12    13  
0    396.90  4.98  24.0  
1    396.90  9.14  21.6  
2    392.83  4.03  34.7  
3    394.63  2.94  33.4  
4    396.90  5.33  36.2  
..      ...   ...   ...  
501  391.99  9.67  22.4  
502  396.90  9.08  20.6  
503  396.90  5.64  23.9  
504  393.45  6.48  22.0  
505  396.90  7.88  11.9  

[506 rows x 14 columns]

In [49]:
# Añado el nombre de las columnas
df.columns =['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

El fichero contiene un conjunto de datos sobre los precios de las viviendas en Boston.
Los datos se muestran ordenados,ces decir que cada dato corresponde a una columna/fila. 

Voy a describir las columnas:

- CRIM: Tasa de criminalidad per cápita por área.
- ZN: Proporción de terreno residencial dividido en zonas para lotes de más de 25,000 pies cuadrados.
- INDUS: Proporción de acres de negocios no minoristas por ciudad.
- CHAS: Variable ficticia del río Charles (1 si el tramo limita con el río; 0 en caso contrario).
- NOX: Concentración de óxidos de nitrógeno (partes por 10 millones).
- RM: Número promedio de habitaciones por vivienda.
- AGE: Proporción de unidades ocupadas por sus propietarios construidas antes de 1940.
- DIS: Distancias ponderadas a cinco centros de empleo de Boston.
- RAD: Índice de accesibilidad a carreteras radiales.
- TAX: Tasa de impuesto a la propiedad por $10,000.
- PTRATIO: Proporción alumno-profesor en la ciudad.
- B: 1000(Bk - 0.63)^2 donde Bk es la proporción de personas de ascendencia afroamericana por ciudad.
- LSTAT: Porcentaje de población de estatus bajo.
- MEDV: Valor medio de las viviendas ocupadas por sus propietarios en miles de dólares.

Voy a mirar el tipo de datos que tenemos:

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


Los tipos de datos son números decimales para todas las columnas, excepto para la columna 'CHAS' y 'RAD' que son variables enteras. 

Ahora, voy a identificar anomalías o datos faltantes:

In [51]:
# Voy a comprobar el tamaño de la base de datos

print('El tamaño de df es:', df.shape)

El tamaño de df es: (506, 14)


La comprobación del tamaño corresponde con lo que indica al mostrar el fichero df, entonces nos indica que no tenemos datos faltantes, ya que coinciden los datos de entrada con tamaño. 
Ahora para identificar anomalías, voy a calcular estadísticas de resumen. 

In [52]:
df.describe().round(2) #Redondeo a 2 decimales 

CRIM      ZN   INDUS    CHAS     NOX      RM     AGE     DIS     RAD  \
count  506.00  506.00  506.00  506.00  506.00  506.00  506.00  506.00  506.00   
mean     3.61   11.36   11.14    0.07    0.55    6.28   68.57    3.80    9.55   
std      8.60   23.32    6.86    0.25    0.12    0.70   28.15    2.11    8.71   
min      0.01    0.00    0.46    0.00    0.38    3.56    2.90    1.13    1.00   
25%      0.08    0.00    5.19    0.00    0.45    5.89   45.02    2.10    4.00   
50%      0.26    0.00    9.69    0.00    0.54    6.21   77.50    3.21    5.00   
75%      3.68   12.50   18.10    0.00    0.62    6.62   94.07    5.19   24.00   
max     88.98  100.00   27.74    1.00    0.87    8.78  100.00   12.13   24.00   

          TAX  PTRATIO       B   LSTAT    MEDV  
count  506.00   506.00  506.00  506.00  506.00  
mean   408.24    18.46  356.67   12.65   22.53  
std    168.54     2.16   91.29    7.14    9.20  
min    187.00    12.60    0.32    1.73    5.00  
25%    279.00    17.40  375.38    6.95   17.02  
50%    330.00    19.05  391.44   11.36   21.20  
75%    666.00    20.20  396.22   16.96   25.00  
max    711.00    22.00  396.90   37.97   50.00

Podemos comprobar que salen datos correctos excepto en la columna de 'ZN' y en la columna 'CHAS', dónde los valores desde el min hasta el tercer quantil son 0. Realmente lo que me indica, es que en el caso de 'ZN' es posible que indique el área no destinada a ese próposito (es decir, tierras residenciales divididas en zonas) y en el caso de 'CHAS', indica las propiedades no están cerca del río. 

### - _Exercici 1_
##### Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

PASO 1. Voy a importar librerias y después establezco dos variables una dependiente que es la columna 'MEDV' y el resto en independientes. 

In [53]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import seaborn as sns

In [54]:
x = df.drop('MEDV', axis=1) # Serían las características
y = df['MEDV'] # Variable dependiente valor medio de las viviendas.

Ahora voy a dividir los datos en train y test 

In [55]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

PASO 2. Voy a comprobar el tamaño de los datos y también mostraré los datos de entrenamiento. 

In [56]:
print('Entrenamiento tamaño características:', x_train.shape)
print('Entrenamiento tamaño etiquetas:', y_train.shape)
print('Test tamaño características:', x_test.shape)
print('Test tamaño etiquetas:', y_test.shape)

Entrenamiento tamaño características: (404, 13)
Entrenamiento tamaño etiquetas: (404,)
Test tamaño características: (102, 13)
Test tamaño etiquetas: (102,)


In [57]:
x_train.head()

CRIM    ZN  INDUS  CHAS     NOX     RM   AGE     DIS  RAD    TAX  \
477  15.02340   0.0  18.10     0  0.6140  5.304  97.3  2.1007   24  666.0   
15    0.62739   0.0   8.14     0  0.5380  5.834  56.5  4.4986    4  307.0   
332   0.03466  35.0   6.06     0  0.4379  6.031  23.3  6.6407    1  304.0   
423   7.05042   0.0  18.10     0  0.6140  6.103  85.1  2.0218   24  666.0   
19    0.72580   0.0   8.14     0  0.5380  5.727  69.5  3.7965    4  307.0   

     PTRATIO       B  LSTAT  
477     20.2  349.48  24.91  
15      21.0  395.62   8.47  
332     16.9  362.25   7.83  
423     20.2    2.52  23.29  
19      21.0  390.95  11.28

In [58]:
y_train.head()

477    12.0
15     19.9
332    19.4
423    13.4
19     18.2
Name: MEDV, dtype: float64

Ahora creo un modelo de regresión lineal, con una sola variable independiente 'MEDV'. 

##### REGRESION LINEAL

In [59]:
# PASO 3. Creo el modelo.
model = LinearRegression()

In [60]:
# Y ahora lo uso en el entrenamiento. 
model.fit(x_train, y_train)

LinearRegression()

In [61]:
# PASO 4. Ahora obtengo resultados
r_sq = model.score(x_train, y_train)
print(f"coeficiente de determinación: {r_sq}")

coeficiente de determinación: 0.7508856358979673


In [62]:
# PASO 5. Predecir respuesta con los datos de x_test
y_pred = model.predict(x_test)
print(f"predecir respuesta:\n{y_pred}")

predecir respuesta:
[28.99672362 36.02556534 14.81694405 25.03197915 18.76987992 23.25442929
 17.66253818 14.34119    23.01320703 20.63245597 24.90850512 18.63883645
 -6.08842184 21.75834668 19.23922576 26.19319733 20.64773313  5.79472718
 40.50033966 17.61289074 27.24909479 30.06625441 11.34179277 24.16077616
 17.86058499 15.83609765 22.78148106 14.57704449 22.43626052 19.19631835
 22.43383455 25.21979081 25.93909562 17.70162434 16.76911711 16.95125411
 31.23340153 20.13246729 23.76579011 24.6322925  13.94204955 32.25576301
 42.67251161 17.32745046 27.27618614 16.99310991 14.07009109 25.90341861
 20.29485982 29.95339638 21.28860173 34.34451856 16.04739105 26.22562412
 39.53939798 22.57950697 18.84531367 32.72531661 25.0673037  12.88628956
 22.68221908 30.48287757 31.52626806 15.90148607 20.22094826 16.71089812
 20.52384893 25.96356264 30.61607978 11.59783023 20.51232627 27.48111878
 11.01962332 15.68096344 23.79316251  6.19929359 21.6039073  41.41377225
 18.76548695  8.87931901 20.830

In [63]:
# Obtención de la respuesta predicha
y_predicha = model.intercept_ + model.coef_ * x_test
print(f"predecir respuesta:\n{y_predicha}")

predecir respuesta:
          CRIM         ZN      INDUS       CHAS        NOX         RM  \
173  30.236375  30.246751  30.410293  30.246751  21.473408  58.726318   
274  30.240370  31.451170  30.505591  33.031189  22.557174  60.244399   
491  30.234796  30.246751  31.366912  30.246751  19.770347  56.804302   
72   30.236391  30.246751  30.683267  30.246751  23.142063  57.168286   
452  29.671277  30.246751  30.977642  30.246751  17.981273  58.198096   
..         ...        ...        ...        ...        ...        ...   
412  28.120056  30.246751  30.977642  30.246751  19.976779  50.789680   
436  28.616394  30.246751  30.977642  30.246751  17.516802  58.926065   
411  28.658236  30.246751  30.977642  30.246751  19.976779  59.796077   
86   30.240886  30.246751  30.428060  30.246751  22.522769  56.946345   
75   30.235997  30.246751  30.764836  30.246751  22.729200  58.149269   

           AGE        DIS        RAD        TAX    PTRATIO          B  \
173  29.717227  26.415265  31.

Para la segunda regresión crearé un RandomForest 

##### RANDOM FOREST

In [64]:
 from sklearn.ensemble import RandomForestRegressor

In [65]:
# PASO 3. Modelo de entrenamiento.
# Creo el modelo con 1000 arboles de decisión.
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
# Entreno el modelo
rf.fit(x_train, y_train)

RandomForestRegressor(n_estimators=1000, random_state=42)

In [66]:
# Usamos los datos de test para las predicciones
predictions = rf.predict(x_test)
print(f"predecir respuesta:\n{predictions}")

predecir respuesta:
[22.8929 30.832  16.9387 23.4999 16.7633 21.314  19.4238 15.5452 21.1557
 20.9295 20.0174 19.6002  8.6537 21.5041 19.2969 25.9516 19.2831  8.3532
 45.7314 14.9231 24.7412 23.8844 14.646  23.8262 14.7581 14.8449 21.2888
 13.9045 19.4787 20.9983 19.7249 23.2369 31.4506 20.2608 14.6167 15.8104
 34.3984 19.2269 20.9753 24.3122 19.0664 29.5378 45.6189 19.3335 22.334
 13.7426 15.2194 24.4927 18.4645 28.3156 21.1065 33.8044 16.7788 26.2689
 45.3449 21.9525 15.6067 32.2191 22.3477 20.5784 25.5803 34.0846 28.9109
 18.4435 26.9301 17.2001 13.872  23.0504 28.5866 15.7227 20.6554 28.3465
 10.1231 21.3818 22.0312  7.1047 20.0738 45.9498 10.9736 12.8983 21.3169
 12.879  19.8869  9.2583 20.4362 26.994  15.405  23.2867 23.5485 17.8092
 21.6547  8.1135 19.4016 18.8734 23.1591 19.7717 40.1955 12.5851 12.4702
 13.2325 20.3092 23.7541]


In [67]:
# Calculamos los errores absolutos
errors = abs(predictions - y_test)
# Imprimir el error absoluto medio (mae)
print('Error absoluto medio:', round(np.mean(errors), 2), 'grados.')

Error absoluto medio: 2.09 grados.


In [68]:
# determinar métricas de rendimiento 
# Calcular la media absoluta del porcentaje de error (MAPE)
mape = 100 * (errors / y_test)
# Calculate and display accuracy
exactitud = 100 - np.mean(mape)
print('Exactitud:', round(exactitud, 2), '%.')

Exactitud: 88.65 %.


### - Exercici 2
#### Compara’ls en base al MSE i al R2.

Voy a calcular el error cuadrático medio i el coeficiente de determinación para los dos modelos. 

In [69]:
from sklearn.metrics import mean_squared_error  
from sklearn.metrics import r2_score   

In [70]:
# Primero calculo MSE para regresión lineal
mse_linear = mean_squared_error(y_test, y_pred) 
# Y ahora el Coeficiente de determinación
R2_linear = r2_score(y_test, y_pred)  


print('MSE:',mse_linear)
print('R2:',R2_linear )

MSE: 24.29111947497371
R2: 0.6687594935356294


In [71]:
# Primero calculo MSE para randomForest
mse_rf = mean_squared_error(y_test, predictions)
# Y ahora el Coeficiente de determinación
R2_linear = r2_score(y_test, predictions)  


print('MSE:',mse_rf)
print('R2:',R2_linear )

MSE: 8.84665565990191
R2: 0.8793645264344092


Viendo los datos se observa que el que da menos error es el RandomForest en relación con la regresión lineal. Y el R2 más alto también es el RandomForest, con lo que es más bueno el segundo método ya que se aproxima más a el valor 1. 

### - Exercici 3
#### Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció

Voy a aplicar Ridge porque sospecho que muchas características tienen cierta relevancia y se busca reducir el sobreajuste.

In [72]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.linear_model import RidgeCV

In [73]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_regression
from sklearn.metrics import mean_squared_error

### Regresión Lineal con diferentes parámetros:

In [74]:
regresion_l = LinearRegression()
regresion_l.get_params()

{'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'positive': False}

In [75]:
# Define los parámetros que quieres probar
param_grid = {
    'alpha' : [0.001, 0.01, 0.1, 10, 100, 1000],
    'fit_intercept': [True, False]
    
}

# Creo un nuevo modelo Ridge
ridge_model = Ridge()
# Realizo la búsqueda de los mejores parámetros
grid_search = GridSearchCV(estimator=ridge_model, param_grid=param_grid, cv=5)
grid_result = grid_search.fit(x_train, y_train)

# Mejores parámetros
best_model_rl = grid_result.best_estimator_

# Calculo el error cuadrático medio para evaluar el modelo
mse = mean_squared_error(y_test, best_model_rl.predict(x_test))
print("Error Cuadrático Medio (Regresión Lineal):", mse)

Error Cuadrático Medio (Regresión Lineal): 24.291177851311257


- En un primer momento sólo configuré el parametro fit_intercept, y no resultó ninguna mejora, entonces voy a añadir alpha con los parámetros de 0.1, 1, 10, 100, y ver que tal va. 
- Me salía otro error de que alpha no estaba en parámetros, así que lo tenía que hacer de otro modo. 
- He tenido que extender el alpha para conseguir que bajara el error cuadrático. 

### Random Forest con diferentes parámetros

In [76]:
arbol = RandomForestRegressor()
arbol.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'squared_error',
 'max_depth': None,
 'max_features': 1.0,
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [77]:
# Definición de parámetros que quieres probar
param_grid_rf = {
    'n_estimators': [10, 20, 30],
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
    
}


# Realizo la búsqueda de los mejores parámetros
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, cv=5)
grid_result_rf = grid_search_rf.fit(x_train, y_train)

# Mejores parámetros
best_model_rfr = grid_result_rf.best_estimator_

# Calculo el error cuadrático medio para evaluar el modelo
mse_rf = mean_squared_error(y_test, best_model_rfr.predict(x_test))

print("Error Cuadrático Medio (Random Forest):", mse_rf)

Error Cuadrático Medio (Random Forest): 8.26004090198577


    Se puede comprobar que en el Random Forest ha bajado el error cuadrático aplicando Ridge. 

### - Exercici 4
#### Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).



In [78]:
# Entrenamiento y evaluación con enfoque a todos los datos para modelo lineal.
modelo_lineal = LinearRegression()
modelo_lineal.fit(x, y)
y_pred_all = modelo_lineal.predict(x)

In [79]:
# Ahora calculo el error cuadrátrico
mse_lineal_all = mean_squared_error(y, y_pred_all)

print("Error Cuadrático Medio (Modelo lineal):", mse_lineal_all)

Error Cuadrático Medio (Modelo lineal): 21.894831181729202


In [80]:
# Entrenamiento y evaluación con enfoque a todos los datos para modelo random forest.
modelo_random = RandomForestRegressor()
modelo_random.fit(x, y)
y_pred_all_rf = modelo_random.predict(x)

In [81]:
# Ahora calculo el error cuadrátrico
mse_random_all = mean_squared_error(y, y_pred_all_rf)

print("Error Cuadrático Medio (Modelo Random Forest):", mse_random_all)

Error Cuadrático Medio (Modelo Random Forest): 1.4514677628458499


Es más que evidente la reducción del error cuadrático de los dos modelos. 

### - Exercici 5
#### No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.



In [82]:
x.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [83]:
# Eliminar la variable 'RM' del nombre de habitaciones. 
x = x.drop('RM', axis=1)

In [84]:
x.head()

CRIM    ZN  INDUS  CHAS    NOX   AGE     DIS  RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  65.2  4.0900    1  296.0     15.3   
1  0.02731   0.0   7.07     0  0.469  78.9  4.9671    2  242.0     17.8   
2  0.02729   0.0   7.07     0  0.469  61.1  4.9671    2  242.0     17.8   
3  0.03237   0.0   2.18     0  0.458  45.8  6.0622    3  222.0     18.7   
4  0.06905   0.0   2.18     0  0.458  54.2  6.0622    3  222.0     18.7   

        B  LSTAT  
0  396.90   4.98  
1  396.90   9.14  
2  392.83   4.03  
3  394.63   2.94  
4  396.90   5.33

In [85]:
# Vuelvo a dividir los datos en train/test.
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [86]:
# Compruebo el tamaño de los datos. 
print('Entrenamiento tamaño características:', x_train.shape)
print('Entrenamiento tamaño etiquetas:', y_train.shape)
print('Test tamaño características:', x_test.shape)
print('Test tamaño etiquetas:', y_test.shape)

Entrenamiento tamaño características: (404, 12)
Entrenamiento tamaño etiquetas: (404,)
Test tamaño características: (102, 12)
Test tamaño etiquetas: (102,)


##### REGRESION LINEAL 

In [87]:
modelo_lineal = LinearRegression()
modelo_lineal.fit(x_train, y_train)

LinearRegression()

In [88]:
# Utilizamos test para predecir 
predict_lineal = modelo_lineal.predict(x_test)

In [89]:
# Calculo el error cuadrático. 
lineal_mse = mean_squared_error(y_test, predict_lineal)

print("Error Cuadrático Medio (Modelo lineal):", lineal_mse)

Error Cuadrático Medio (Modelo lineal): 22.133246888189614


En relación con los anteriores MSE de la regresión lineal; 
- En la primera aplicando la regresión era: 24.29111947497371
- En la segunda aplicando el ridge a la regresión era: 24.291177851311257
- En la tercera valorando todos los datos: 21.894831181729202
- En la cuarta quitando la variable 'RM': 22.133246888189614

En el segundo proceso, me ha costado mucho bajarla pero en el tercero valorando los datos sí que se ha conseguido bajar. Finalmente en el tercero, puedo decir que esa variable perjudica dentro de los datos, y sin ella el modelo de precisión es más conciso, aunque no existe mucha diferencia con la anterior consulta. 

#### RANDOM FOREST REGRESSOR

In [90]:
modelo_random_forest = RandomForestRegressor()
modelo_random_forest.fit(x_train, y_train)

RandomForestRegressor()

In [91]:
# Utilizamos test para predecir 
predict_rf = modelo_random_forest.predict(x_test)

In [92]:
# Calculo el error cuadrático. 
random_mse = mean_squared_error(y_test, predict_rf)

print("Error Cuadrático Medio (Modelo lineal):", random_mse)

Error Cuadrático Medio (Modelo lineal): 14.536730558823523


En relación con los anteriores MSE del modelo Random Forest;

- En la primera aplicando la Random Forest era: 8.84665565990191
- En la segunda aplicando el ridge a la Random Forest era: 8.26004090198577
- En la tercera valorando todos los datos del conjunto: 1.4260920296442683
- En la cuarta quitando la variable 'RM': 14.956887676470577

Bien, en este caso, solo aplicando el método Random Forest el error era muy pequeño, después al aplicar el proceso de Ridge, se ha conseguido bajar un poquito, pero en la valoración de todos los datos del conjunto realmente ha sido un éxito ya que el error cuadrático rondaba el 1.5. Finalmente al quitar la variable de RM, de nuevo a subido, así que no ha mejorado los datos en este modelo. 
